src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ResourceProtoSerializer.cs

```diff
@@ -0,0 +1,58 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.Resources;
+
+namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation.Serializer;
+
+internal static class ResourceProtoSerializer
+{
+    private const int ReserveSizeForLength = 4;
+
+    internal static int WriteResource(byte[] buffer, int writePosition, Resource? resource)
+    {
+        OtlpProtoTagWriter.OtlpTagWriterState otlpTagWriterState = new OtlpProtoTagWriter.OtlpTagWriterState
+        {
+            Buffer = buffer,
+            WritePosition = writePosition,
+        };
+
+        if (resource != null && resource != Resource.Empty)
+        {
+            otlpTagWriterState.WritePosition = ProtobufSerializer.WriteTag(otlpTagWriterState.Buffer, otlpTagWriterState.WritePosition, FieldNumberConstants.ResourceSpans_Resource, ProtobufWireType.LEN);
+            int resourceLengthPosition = otlpTagWriterState.WritePosition;
+            otlpTagWriterState.WritePosition += ReserveSizeForLength;
+
+            foreach (KeyValuePair<string, object> attribute in resource.Attributes)
+            {
+                otlpTagWriterState = ProcessResourceAttribute(otlpTagWriterState, attribute);
+            }
+
+            if (!resource.Attributes.Any(kvp => kvp.Key == ResourceSemanticConventions.AttributeServiceName))
+            {
+                var serviceName = (string)ResourceBuilder.CreateDefault().Build().Attributes.FirstOrDefault(
+                    kvp => kvp.Key == ResourceSemanticConventions.AttributeServiceName).Value;
+
+                otlpTagWriterState = ProcessResourceAttribute(otlpTagWriterState, new KeyValuePair<string, object>(ResourceSemanticConventions.AttributeServiceName, serviceName));
+            }
+
+            var resourceLength = otlpTagWriterState.WritePosition - (resourceLengthPosition + ReserveSizeForLength);
+            ProtobufSerializer.WriteReservedLength(otlpTagWriterState.Buffer, resourceLengthPosition, resourceLength);
+        }
+
+        return otlpTagWriterState.WritePosition;
+    }
+
+    private static OtlpProtoTagWriter.OtlpTagWriterState ProcessResourceAttribute(OtlpProtoTagWriter.OtlpTagWriterState otlpTagWriterState, KeyValuePair<string, object> attribute)
```

Comment:
Is this working? I think you need to pass the state by ref otherwise in the body here you are modifying a copy.

```diff
 private static OtlpProtoTagWriter.OtlpTagWriterState ProcessResourceAttribute(
-   OtlpProtoTagWriter.OtlpTagWriterState otlpTagWriterState,
+   ref OtlpProtoTagWriter.OtlpTagWriterState otlpTagWriterState,
    KeyValuePair<string, object> attribute)
```